### PR TITLE
Add an option to turn segfaults into julia errors

### PR DESCRIPTION
**Pull request by Keno**
This is quite handy when developing against C or C++ libraries that
like to segfault on invalid parameters. I've been carrying this patch locally,
but it's quite useful, so I wanted to make it available.

### CODE REVIEW

**Comment by garrison** [comment]
Very cool!

**Comment by kmsquire** [comment]
:100:

File: src/init.c [hunk 1]
```diff
@@ -289,9 +289,18 @@ void segv_handler(int sig, siginfo_t *info, void *context)
         sigprocmask(SIG_UNBLOCK, &sset, NULL);
         jl_throw(jl_memory_exception);
     }
+#ifdef SEGV_EXCEPTION
+    else {
```
Comment (vtjnash) [inline src/init.c]: This should be conditional on sig actually being a SEGV

**Comment by vtjnash** [comment]
Can you merge this with the "in_jl_" code and remove the conditional compile?

File: src/init.c (hunk 1 above)
Comment (Keno) [inline src/init.c]: Maybe, but I'd really like to catch everything, whether it's SIGSEGV or SIGBUS (not sure what else gets trapped here). I see that the julia side error name is then misleading, but I think that's fine.

**Comment by Keno** [comment]
Sure that sounds reasonable.

File: src/init.c (hunk 1 above)
Comment (vtjnash) [inline src/init.c]: SIGILL, SIGTINFO, SIGUSR1, SIGKILL I believe all go through here also. The last three should not be thrown as Julia errors

**Comment by ScottPJones** [comment]
Has this been fixed to not catch everything (it would be pretty bad if it trapped SIGKILL, otherwise how do you stop a run-away process)?

**Comment by vtjnash** [comment]
Unix stops running code and cleans everything up a fraction of a second after delivering SIGKILL (generally seems to be ~0.1s)

**Comment by ScottPJones** [comment]
OK, I'm used to having my code handle some cleanup work when a SIGKILL is received... not try to put out a backtrace or something...

**Comment by aviks** [comment]
@ScottPJones  Was there a hint of sarcasm in your comment? Not sure if that was intended... In any case, I believe this is a opt in feature meant to be used for development of the internals, for which you need to recompile Julia from source.

**Comment by ScottPJones** [comment]
No! No sarcasm at all! (sheez, what you guys assume about me! :grinning: I know I am a big PITA, but I really am not trying to pick on anybody or anything! If you talked to me in person, you'd see what a fanboy I am of all the Julia contributors! :blush:)

**Pull request merged**